rust/kernel/net/device.rs

```diff
@@ -0,0 +1,957 @@
+    pub fn get_priv_data(&self) -> &T {
+        // SAFETY: self.ptr is valid if self is valid
+        let priv_ptr = unsafe { rust_helper_netdev_priv(self.ptr) } as *mut T;
+
+        // SAFETY: ptr is valid and of type T
+        unsafe { priv_ptr.as_ref() }.unwrap()
+    }
+
+    /// Return a mutable refference of the private data of the [`NetDevice`]
+    pub fn get_priv_data_mut(&mut self) -> &mut T {
+        // SAFETY: self.ptr is valid if self is valid
+        let priv_ptr = unsafe { rust_helper_netdev_priv(self.ptr) as *mut T };
+
+        // SAFETY: ptr is valid and of type T
+        unsafe { priv_ptr.as_mut().unwrap() }
+    }
+
+    /// ether_setup - setup Ethernet network device
+    ///
+    /// Fill in the fields of the device structure with Ethernet-generic values.
+    pub fn ether_setup(&mut self) {
+        // SAFETY: self.ptr is valid
+        unsafe { bindings::ether_setup(self.ptr as *mut bindings::net_device) }
+    }
+
+    /// hw_addr_random - Generate software assigned random Ethernet and set device flag
+    ///
+    /// Generate a random Ethernet address (MAC) to be used by a net device
+    /// and set addr_assign_type so the state can be read by sysfs and be
+    /// used by userspace.
+    pub fn hw_addr_random(&mut self) {
+        // SAFETY: self.ptr is valid
+        unsafe { rust_helper_eth_hw_addr_random(self.ptr) };
+    }
+
+    /// register - register a network device
+    ///
+    /// Take a completed network device structure and add it to the kernel
+    /// interfaces. A %NETDEV_REGISTER message is sent to the netdev notifier
+    /// chain. 0 is returned on success. A negative errno code is returned
+    /// on a failure to set up the device, or if the name is a duplicate.
+    ///
+    /// This is a wrapper around register_netdevice that takes the rtnl semaphore
+    /// and expands the device name if you passed a format string to
+    /// alloc_netdev.
+    pub fn register(&self) -> KernelResult<()> {
+        // SAFETY: self.ptr is valid
+        let err = unsafe { bindings::register_netdev(self.ptr as *mut bindings::net_device) };
+
+        if err != 0 {
+            Err(Error::from_kernel_errno(err))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// register_locked - register a network device if the RtnlLock is already hold
+    ///
+    /// Take a completed network device structure and add it to the kernel
+    /// interfaces. A %NETDEV_REGISTER message is sent to the netdev notifier
+    /// chain. 0 is returned on success. A negative errno code is returned
+    /// on a failure to set up the device, or if the name is a duplicate.
+    ///
+    /// Callers must hold the rtnl semaphore. You may want
+    /// [`register`] instead of this.
+    ///
+    /// BUGS:
+    /// The locking appears insufficient to guarantee two parallel registers
+    /// will not get the same name.
+    ///
+    /// # Safety
+    ///
+    /// caller must hold the RtnlLock and semaphore
```

Comment:
```suggestion
    /// Caller must hold the [`RtnlLock`] and semaphore.
```

And similar everywhere else.